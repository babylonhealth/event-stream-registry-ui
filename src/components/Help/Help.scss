@import "../EventStreamRegistry/_variables.scss";

$largeFontSize: 17px;
$mediumFontSize: 12px;
$smallFontSize: 10px;

.help {
    padding-left: $lpad;
    padding-top: 20px;
    width: 80%;
    font-family: "Courier New", monospace;
    font-size: $smallFontSize;
}

.help p {
    margin: 0;
}

.help li {
    list-style-type: circle;
}

.help h1 {
    margin-top: 0;
    padding-bottom: 1px;
    font-size: $largeFontSize;
    font-weight: lighter;
    border-bottom: 1px lightgray solid;
}

.help h2 {
    font-weight: lighter;
    font-size: $mediumFontSize;
    margin: 0;
    width: 100%;
    border-bottom: 1px $lightGrey solid;
}

.help h3 {
    font-weight: lighter;
    font-size: $mediumFontSize;
}
.box {
    padding-bottom: 35px;
}

.help ul dl {
    display: inline-block;
    width: 200px;
    margin: 1px;
}

.help ul dt {
    display: inline-block;
}
.help ul span {
    font-size: 8px;
}
.help ul p {
    font-size: $mediumFontSize;
}

.fields {
    display: inline-block;
    vertical-align: top;
    width: 45%;
    padding-top: 10px;
}
.node-type {
    display: inline-block;
    vertical-align: top;
    width: 40%;
}
.right {
    float: right;
}
.no-bottom-padding{
    margin-bottom: 0;
}
