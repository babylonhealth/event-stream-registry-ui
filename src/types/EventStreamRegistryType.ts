/**
 * The Event Stream Registry dataset.
 *
 * @param node - contains the set of services and topics in the Event Stream Registry.
 * @param links - defines the connectedness of the nodes.
 * @type {EventStreamRegistryType}
 */
export interface EventStreamRegistryType {
    nodes: EventStreamRegistryNode[],
    links: EventStreamRegistryLink[]
}

/**
 * A single Event Stream Registry node - either a topic or a service.
 *
 * @param id - the node id
 * @param type - the node type, either "SERVICE" or "TOPIC"
 * @param metadata - a map node metadata, the key is the metadata name, the value is a Metadata object
 * @param urls - a list of associated node urls
 * @type {EventStreamRegistryNode}
 * */
export interface EventStreamRegistryNode {
    id: string,
    type: NodeType,
    metadata: Record<string, Metadata>,
    urls: Url[],
}

/**
 * A metadata value.
 *
 * @param value - the metadata value
 * @param type - either 'NUMBER' or 'STRING', the metadata type, the type defines allowed filtering operations
 * @param url? - an associated url, renders values as links
 * @param status? - either 'GOOD' or 'BAD', indicates the "health" of a metric, e.g. a high consumer lag might be
 * consider 'unhealthy` of `BAD`, renders values `green` and `red`.
 * @type {Metadata}
 * */
export interface Metadata {
    "value": string | number,
    "type": MetadataType
    "url": string | undefined
    "status": Status | undefined
}

/**
 * An associated service or topic url.
 *
 * @param name - the name of the url, e.g. grafana
 * @param value - the url, e.g. https://my.grafana.co.uk/event-stream-dashboard/my-topic
 * @param icon? - a material ui icon to render urls, if null urls are rendered as 'launch' icon
 * @type {Url}
 * */
export interface Url {
    "name": string,
    "value": string,
    "icon": string | undefined
}

/**
 * Defines a link between two nodes
 *
 * @param source - the node id of the source
 * @param target - the node id of the target
 * @type {EventStreamRegistryLink}
 * */
export interface EventStreamRegistryLink {
    source: string,
    target: string,
}

export enum MetadataType {
    "NUMBER" = "NUMBER",
    "STRING" = "STRING"
}

export enum Status {
    "BAD",
    "GOOD",
}

export enum StatusColors {
    BAD = "#FF0000",
    GOOD = "#008000",
    NO_STATUS = "#000000",
}

export enum NodeType {
    "TOPIC" = "TOPIC",
    "SERVICE" = "SERVICE"
}
