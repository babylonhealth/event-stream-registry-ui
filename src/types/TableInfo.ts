export type MetadataFields = {
    numeric: Set<string>,
    string: Set<string>,
    url: Set<string>
}

export type EventStreamRegistryMetadata = {
    topic: MetadataFields,
    service: MetadataFields
}
export type QueryParameter = {
    key: string,
    value: string
    operation: string
}
export enum View {
    "GRAPH"= "GRAPH",
    "SERVICES"= "SERVICES",
    "TOPICS"= "TOPICS",
    "HELP"= "HELP"
}

