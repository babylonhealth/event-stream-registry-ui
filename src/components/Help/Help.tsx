import * as React from "react";

import {EventStreamRegistryMetadata} from "../../types/TableInfo";
import "./Help.scss"
import {ReactNode} from "react";

type HelpProps = {
    metadataInfo: EventStreamRegistryMetadata,
}

type BulletPointItemProps = {
    title: ReactNode,
    desc: ReactNode,
}

type CompositionProps = {
    children: ReactNode,
}


const BulletPointItem = ({title, desc}: BulletPointItemProps) => {
    return (
        <li>
            <dl>{title}</dl>
            <dt>{desc}
            </dt>
        </li>
    )
};

const BulletPointList = ({children}: CompositionProps) => {
    return (
        <ul>
            {children}
        </ul>
    )
};

const HelpContainer = ({children}: CompositionProps) => {
    return (
        <div className={"box"}>
            {children}
        </div>
    )
};

const Help = ({metadataInfo}: HelpProps) => {
    return (
        <div className={"help"}>

            <HelpContainer>
                <p>The Event Stream Registry UI is a tool for observing and monitoring event-streams.</p>
            </HelpContainer>

            <HelpContainer>
                <h1>VIEW COMMANDS</h1>
                <p>There are four views of the Event Stream Registry UI</p>
                <BulletPointList>
                    <BulletPointItem
                        title={<>GRAPH</>}
                        desc={<>Node graph diagram showing the connectedness of services and topics in the Event
                            Registry</>}/>
                    <BulletPointItem
                        title={<>TOPICS</>}
                        desc={<>Table of Event Stream Registry topic information</>}/>
                    <BulletPointItem
                        title={<>SERVICES</>}
                        desc={<>Table of Event Stream Registry service information</>}/>
                    <BulletPointItem
                        title={<>HELP</>}
                        desc={<>Information on how to use the Event Stream Registry</>}/>
                </BulletPointList>
                <p>You can switch view using the search bar by typing `topics`, `services`, `graph` and `help` or the buttons in the top right corner.</p>
            </HelpContainer>

            <HelpContainer>
                <h1>SEARCH PARAMETERS</h1>
                <p>The search bar can be used to filter <b>TOPICS</b> and <b>SERVICES</b>.</p>
                <BulletPointList>
                    <BulletPointItem
                        title={<>status=bad</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> in an unhealthy state</>}/>
                    <BulletPointItem
                        title={<>status=good</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> in a healthy state</>}/>
                    <BulletPointItem
                        title={<>id=dfh</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> where id is not null and id contains <i>"dfh"</i></>}/>
                    <BulletPointItem
                        title={<>partitions=6</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> where partitions is not null and partitions equals
                            6</>}/>
                    <BulletPointItem
                        title={<>offset&gt;100</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> where offset is not null and offset is greater
                            than 100</>}/>
                </BulletPointList>
                <p>You can also filter by multiple fields. When filtering by multiple fields the operations are
                    conjugated.</p>
                <BulletPointList>
                    <BulletPointItem
                        title={<>offset&gt;100 partition=6</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> where offset is not null and offset is greater
                            than 100 AND where partitions is not null and partitions is equal to 6 </>}/>
                    <BulletPointItem
                        title={<>id=dfh status=bad</>}
                        desc={<>Get <b>TOPICS</b> and <b>SERVICES</b> in an unhealthy state <b>AND</b> where id is not
                            null and id contains "dfh"</>}/>
                </BulletPointList>
                <p>The following operations are available for each field type:</p>
                <BulletPointList>
                    <BulletPointItem
                        title={<>NUMERIC FIELD</>}
                        desc={<> <b>=
                            &nbsp;&nbsp;&lt;
                            &nbsp;&nbsp;&lt;=
                            &nbsp;&nbsp;&gt;
                            &nbsp;&nbsp;&gt;=
                        </b></>}/>
                    <BulletPointItem
                        title={<>STRING FIELD</>}
                        desc={<><b>=</b>
                            &nbsp;&nbsp;&nbsp;
                            <span><i>evaluted as contains, i.e. id=abc equates to the string `abc` is contained within
                                the id</i></span></>}/>
                </BulletPointList>
                <p>You can only filter by known fields. The list of known fields is shown below. </p>
            </HelpContainer>

            <HelpContainer>
                <div className={"fields"}>
                    <h1>TOPIC FIELDS</h1>
                    <div className={"node-type"}>
                        <h3>NUMERIC</h3>
                        <BulletPointList>
                            {Array.from(metadataInfo.topic.numeric).map((field, i) => <li key={i}
                                                                                          className={"field"}>{field}</li>)}
                        </BulletPointList>
                    </div>
                    <div className={"node-type left"}>
                        <h3>STRING</h3>
                        <BulletPointList>
                            <li className={"field"}>{"id"}</li>
                            {Array.from(metadataInfo.topic.string).map((field, i) =>
                                <li key={i}className={"field"}>{field}</li>)}
                        </BulletPointList>
                    </div>
                </div>
                <div className={"fields right"}>
                    <h1>SERVICE FIELDS</h1>
                    <div className={"node-type"}>
                        <h3>NUMERIC</h3>
                        <BulletPointList>
                            {Array.from(metadataInfo.service.numeric).map((field, i) =>
                                <li key={i} className={"field"}>{field}</li>)}
                        </BulletPointList>
                    </div>
                    <div className={"node-type"}>
                        <h3>STRING</h3>
                        <BulletPointList>
                            <li>{"id"}</li>
                            {Array.from(metadataInfo.service.string).map((field, i) => <li key={i}
                                                                                           className={"field"}>{field}</li>)}
                        </BulletPointList>
                    </div>
                </div>
            </HelpContainer>
        </div>
    )
};

export default Help;
