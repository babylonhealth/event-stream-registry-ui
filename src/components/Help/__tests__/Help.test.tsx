import React from 'react'
import {cleanup, render} from '@testing-library/react'
import Help from '../Help'
import {EventStreamRegistryMetadata} from "../../../types/TableInfo";

afterEach(cleanup);

it('Help snapshot test', () => {
    const metadataInfo: EventStreamRegistryMetadata = {
        "topic": {
            "numeric": new Set(["t_num_1", "t_num_2"]),
            "string": new Set(["t_str_1"]),
            "url": new Set(["u_1"]),
        },
        "service": {
            "numeric": new Set(["s_num_1", "s_num_2"]),
            "string": new Set(),
            "url": new Set(),
        },
    };

    const {asFragment} = render(<Help metadataInfo={metadataInfo}/>);

    // @ts-ignore
    expect(asFragment(<Help />)).toMatchSnapshot()
});
