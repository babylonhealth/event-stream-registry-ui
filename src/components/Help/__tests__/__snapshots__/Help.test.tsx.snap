// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Help snapshot test 1`] = `
<DocumentFragment>
  <div
    class="help"
  >
    <div
      class="box"
    >
      <p>
        The Event Stream Registry UI is a tool for observing and monitoring event-streams.
      </p>
    </div>
    <div
      class="box"
    >
      <h1>
        VIEW COMMANDS
      </h1>
      <p>
        There are four views of the Event Stream Registry UI
      </p>
      <ul>
        <li>
          <dl>
            GRAPH
          </dl>
          <dt>
            Node graph diagram showing the connectedness of services and topics in the Event Registry
          </dt>
        </li>
        <li>
          <dl>
            TOPICS
          </dl>
          <dt>
            Table of Event Stream Registry topic information
          </dt>
        </li>
        <li>
          <dl>
            SERVICES
          </dl>
          <dt>
            Table of Event Stream Registry service information
          </dt>
        </li>
        <li>
          <dl>
            HELP
          </dl>
          <dt>
            Information on how to use the Event Stream Registry
          </dt>
        </li>
      </ul>
      <p>
        You can switch view using the search bar by typing \`topics\`, \`services\`, \`graph\` and \`help\` or the buttons in the top right corner.
      </p>
    </div>
    <div
      class="box"
    >
      <h1>
        SEARCH PARAMETERS
      </h1>
      <p>
        The search bar can be used to filter 
        <b>
          TOPICS
        </b>
         and 
        <b>
          SERVICES
        </b>
        .
      </p>
      <ul>
        <li>
          <dl>
            status=bad
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             in an unhealthy state
          </dt>
        </li>
        <li>
          <dl>
            status=good
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             in a healthy state
          </dt>
        </li>
        <li>
          <dl>
            id=dfh
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             where id is not null and id contains 
            <i>
              "dfh"
            </i>
          </dt>
        </li>
        <li>
          <dl>
            partitions=6
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             where partitions is not null and partitions equals 6
          </dt>
        </li>
        <li>
          <dl>
            offset&gt;100
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             where offset is not null and offset is greater than 100
          </dt>
        </li>
      </ul>
      <p>
        You can also filter by multiple fields. When filtering by multiple fields the operations are conjugated.
      </p>
      <ul>
        <li>
          <dl>
            offset&gt;100 partition=6
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             where offset is not null and offset is greater than 100 AND where partitions is not null and partitions is equal to 6 
          </dt>
        </li>
        <li>
          <dl>
            id=dfh status=bad
          </dl>
          <dt>
            Get 
            <b>
              TOPICS
            </b>
             and 
            <b>
              SERVICES
            </b>
             in an unhealthy state 
            <b>
              AND
            </b>
             where id is not null and id contains "dfh"
          </dt>
        </li>
      </ul>
      <p>
        The following operations are available for each field type:
      </p>
      <ul>
        <li>
          <dl>
            NUMERIC FIELD
          </dl>
          <dt>
             
            <b>
              =   &lt;   &lt;=   &gt;   &gt;=
            </b>
          </dt>
        </li>
        <li>
          <dl>
            STRING FIELD
          </dl>
          <dt>
            <b>
              =
            </b>
               
            <span>
              <i>
                evaluted as contains, i.e. id=abc equates to the string \`abc\` is contained within the id
              </i>
            </span>
          </dt>
        </li>
      </ul>
      <p>
        You can only filter by known fields. The list of known fields is shown below. 
      </p>
    </div>
    <div
      class="box"
    >
      <div
        class="fields"
      >
        <h1>
          TOPIC FIELDS
        </h1>
        <div
          class="node-type"
        >
          <h3>
            NUMERIC
          </h3>
          <ul>
            <li
              class="field"
            >
              t_num_1
            </li>
            <li
              class="field"
            >
              t_num_2
            </li>
          </ul>
        </div>
        <div
          class="node-type left"
        >
          <h3>
            STRING
          </h3>
          <ul>
            <li
              class="field"
            >
              id
            </li>
            <li
              class="field"
            >
              t_str_1
            </li>
          </ul>
        </div>
      </div>
      <div
        class="fields right"
      >
        <h1>
          SERVICE FIELDS
        </h1>
        <div
          class="node-type"
        >
          <h3>
            NUMERIC
          </h3>
          <ul>
            <li
              class="field"
            >
              s_num_1
            </li>
            <li
              class="field"
            >
              s_num_2
            </li>
          </ul>
        </div>
        <div
          class="node-type"
        >
          <h3>
            STRING
          </h3>
          <ul>
            <li>
              id
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
